### PR TITLE
Cleanup diff by re-building an env based on project's metadata

### PR DESCRIPTION
**Pull request by chalasr**

### CODE REVIEW

**Comment by chalasr** [comment]
Diff best viewed [without whitespaces](https://github.com/manala/manalize/pull/61/files?w=1)

File: src/Env/Metadata/MetadataParser.php [hunk 1]
```diff
@@ -28,9 +28,9 @@
      *
      * @return MetadataBag
      */
-    public static function parse(EnvEnum $envType)
+    public static function parse(EnvEnum $envType, $metadataPath = null)
```
Comment (ogizanagi) [inline src/Env/Metadata/MetadataParser.php]: I think it's cleaner to have two methods:
- `parse(string $path)`
- `parseEnv(EnvType $envType)` in which is only build the path and called `parse(string $path)`

Also, `METADATA_PATH` may be renamed to `ENV_METADATA_PATH` or whatever, as there is two different metadata types.

File: src/Handler/Diff.php [hunk 2]
```diff
@@ -126,13 +129,18 @@ public function getErrorOutput()
...
         $this->fs->mkdir($tmpPath);
-        $this->fs->mirror($templatePath, $tmpPath);
-        $this->fs->remove("$tmpPath/manala.yml");
+
+        $env = EnvFactory::createFromMetadata($this->envType, MetadataParser::parse($this->envType, $this->cwd.'/ansible/.manalize.yml'));
+        $dump = Dumper::dump($env, $tmpPath);
+
+        do {
+            $dump->next();
+        } while ($dump->valid());
```
Comment (ogizanagi) [inline src/Handler/Diff.php]:
``` php
for (; $dump->valid(); $dump->next());
```

or: 

``` php
foreach ($dump as $̄); // Yes it works
```

is way more fun 😄 

(See https://twitter.com/nicolasgrekas/status/586108028421804032 about the variable name in the foreach :D )

File: src/Env/Config/Variable/SingleValue.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    public static function createFromReplaces(array $replaces)
+    {
+        return new static($replaces[static::getPlaceholder()]);
+    }
+
+    /**
+     * @return string
+     */
+    abstract protected static function getPlaceholder();
```
Comment (ogizanagi) [inline src/Env/Config/Variable/SingleValue.php]: I don't know if we should really care, but according to http://php.net/manual/fr/migration52.incompatible.php : 

> Dropped abstract static class functions. Due to an oversight, PHP 5.0.x and 5.1.x allowed abstract static functions in classes. As of PHP 5.2.x, only interfaces can have them.

(I have the warning in PHPStorm). 
Maybe this should be moved in an interface ? :/

File: src/Env/Env.php [hunk 4]
```diff
@@ -43,9 +44,21 @@ public function getConfigs()
...
     {
         $vars = [];
-        foreach ($this->getVars() as $var) {
-            foreach ($var->getReplaces() as $placeholder => $value) {
-                $vars[$placeholder] = $value;
+        $dependencies = [];
+
+        foreach ($this->getConfigs() as $config) {
+            foreach ($config->getVars() as $var) {
+                if ($var instanceof Dependency) {
```
Comment (ogizanagi) [inline src/Env/Env.php]: I'm not convinced the `Env` class should be aware of that :/

File: src/Handler/Diff.php [hunk 5]
```diff
@@ -126,13 +129,18 @@ public function getErrorOutput()
...
+
+        $this->fs->remove("$tmpPath/ansible/.manalize.yml");
```
Comment (ogizanagi) [inline src/Handler/Diff.php]: Maybe we should split the env vs metadata dump in order to avoid that.

File: src/Handler/Diff.php (hunk 2 above)
Comment (chalasr) [reply src/Handler/Diff.php]: `$̄` WTF? :joy:

File: src/Env/Dumper.php [hunk 6]
```diff
@@ -25,21 +24,16 @@ class Dumper
...
     {
         $fs = new Filesystem();
 
-        $export = $env->export();
-        $fs->dumpFile("$workDir/ansible/.manalize.yml", Yaml::dump([
-            'envs' => [
-                $export['env'] => ['vars' => $export['vars']],
-            ],
-        ], 5));
+        $fs->dumpFile("$workDir/ansible/.manalize", serialize($env));
```
Comment (ogizanagi) [inline src/Env/Dumper.php]: You should still keeps the multi environments for the future, so this should at least be an array:

``` php
$fs->dumpFile("$workDir/ansible/.manalize", serialize([$env]));
```

Or maybe we don't care right now.

**Comment by chalasr** [comment]
Thanks @ogizanagi!

**Pull request merged**